sanic/server.py

```diff
@@ -99,21 +99,118 @@ def __init__(self, transport: TransportProtocol, unix=None):
             self.client_port = addr[1]
 
 
-class HttpProtocol(asyncio.Protocol):
-    """
-    This class provides a basic HTTP implementation of the sanic framework.
-    """
-
+class SanicProtocol(asyncio.Protocol):
     __slots__ = (
         # app
         "app",
         # event loop, connection
         "loop",
         "transport",
         "connections",
-        "signal",
         "conn_info",
-        "ctx",
+        "signal",
+        "_can_write",
+        "_time",
+        "_task",
+        "_unix",
+        "_data_received",
+    )
+
+    def __init__(
+        self,
+        *,
+        loop,
+        app: Sanic,
+        signal=None,
+        connections=None,
+        unix=None,
+        **kwargs,
+    ):
+        asyncio.set_event_loop(loop)
+        self.loop = loop
+        self.app: Sanic = app
+        self.signal = signal or Signal()
+        self.transport: Optional[Transport] = None
+        self.connections = connections if connections is not None else set()
+        self.conn_info: Optional[ConnInfo] = None
+        self._can_write = asyncio.Event()
+        self._can_write.set()
+        self._unix = unix
+        self._time = 0.0  # type: float
+        self._task = None  # type: Optional[asyncio.Task]
+        self._data_received = asyncio.Event()
+
+    @property
+    def ctx(self):
+        if self.conn_info is not None:
+            return self.conn_info.ctx
+        else:
+            return self.ctx
```

Comment:
RecursionError

Reply:
Yep, thanks. That is a diff left over from a different change I made, not sure if I need it.